### PR TITLE
chore(gatsby): Convert jobs-manager to typescript

### PR DESCRIPTION
**Pull request by martijnjanssen**
## Description
This converts the `gatsby/util/jobs-manager` and the corresponding tests to typescript, one more step towards a fully typed core codebase! I do think that I have gotten the lion's share of the types correct and don't think I have forgotten anything, but if I did, please say so :pray: 

## Related Issues

Related to #21995

### CODE REVIEW

File: packages/gatsby/src/utils/jobs-manager.ts [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
+import uuid from "uuid/v4"
+import path from "path"
+import hasha from "hasha"
+import fs from "fs-extra"
+import pDefer from "p-defer"
+import _ from "lodash"
+import { createContentDigest, slash } from "gatsby-core-utils"
+import reporter from "gatsby-cli/lib/reporter"
 
 const MESSAGE_TYPES = {
```
Comment (blainekasten) [inline packages/gatsby/src/utils/jobs-manager.ts]: can we change this to an enum?

File: packages/gatsby/src/utils/jobs-manager.ts [hunk 2]
```diff
@@ -24,13 +24,22 @@ const jobsInProcess = new Map()
 /** @type {Map<string, {job: InternalJob, deferred: pDefer.DeferredPromise<any>}>} */
 const externalJobsMap = new Map()
```
Comment (blainekasten) [inline packages/gatsby/src/utils/jobs-manager.ts]: can you pass the type to these two maps?  something like:

```ts
new Map<string, {job: InternalJob, pDefer.DeferedPromise<any>}>
```

Comment (martijnjanssen) [reply packages/gatsby/src/utils/jobs-manager.ts]: Yeah, I have some more changes locally, but I've been having some issues with converting the tests for it to typescript. Is looks like tests are failing due to the fact that the imports are different, which `jest.isolateModules` doesn't like. I'll work some more on those changes and push them :)

File: packages/gatsby/src/utils/jobs-manager.ts (hunk 1 above)
Comment (martijnjanssen) [reply packages/gatsby/src/utils/jobs-manager.ts]: Yeah, good idea.

**Comment by martijnjanssen** [comment]
I completely forgot to push my changes. I have reverted the changes to the tests, since I cannot figure out how to handle module isolation in typescript.

**Comment by pvdz** [comment]
(I've converted this PR into a Draft PR)

**Comment by martijnjanssen** [comment]
This is not a draft, the PR is finished in my opinion, just waiting for a final review. I should've removed that label :slightly_smiling_face:

File: packages/gatsby/src/utils/jobs-manager.ts [hunk 3]
```diff
@@ -39,48 +50,60 @@ const convertPathsToAbsolute = filePath => {
...
+  plugin: {
+    name: string
+    version: string
+    resolve: string
+    isLocal: boolean
+  }
+}
+
+export type JobInput = IBaseJob & IJobInput
+export type InternalJob = IBaseJob & IInternalJob
```
Comment (pvdz) [inline packages/gatsby/src/utils/jobs-manager.ts:89]: Please put typing at the top, but after the imports.

File: packages/gatsby/src/utils/jobs-manager.ts [hunk 4]
```diff
@@ -39,48 +50,60 @@ const convertPathsToAbsolute = filePath => {
...
 
-/** @type {pDefer.DeferredPromise<void>|null} */
-let hasActiveJobs = null
+let hasActiveJobs: pDefer.DeferredPromise<void> | null = null
```
Comment (pvdz) [inline packages/gatsby/src/utils/jobs-manager.ts:91]: Can we initialize this with `pDefer<void>().resolve()`? Then in `waitUntilAllJobsComplete` we can unconditionally return `hasActiveJobs.promise` and won't have this slightly awkward nullable property.

This might be considered a semantic change that should be pushed in a separate PR, though.

File: packages/gatsby/src/utils/jobs-manager.ts [hunk 5]
```diff
@@ -39,48 +50,60 @@ const convertPathsToAbsolute = filePath => {
...
+interface IJobInput {
+  inputPaths: string[]
+  plugin: {
+    name: string
+    version: string
+    resolve: string
+  }
+}
+
+interface IInternalJob {
```
Comment (pvdz) [inline packages/gatsby/src/utils/jobs-manager.ts:73]: It's a little next level, but if this had a generic param then you could type the return type of the job and prevent the `Promise<object>` occurrences, but rather give them explicit (even by parametric) types. I'm not sure how much work it would be to get that in, though. So I'm not asking for this change, just something to consider.

File: packages/gatsby/src/utils/jobs-manager.ts [hunk 6]
```diff
@@ -192,17 +211,19 @@ const runJob = (job, forceLocal = false) => {
   }
 }
 
+function isInternalJob(job: JobInput | InternalJob): job is InternalJob {
+  return (job as InternalJob).id !== undefined
```
Comment (pvdz) [inline packages/gatsby/src/utils/jobs-manager.ts:215]: The original code calso checked for `contentDigest`, was that intentionally dropped?

File: packages/gatsby/src/utils/jobs-manager.ts [hunk 7]
```diff
@@ -250,29 +270,26 @@ exports.createInternalJob = (job, plugin) => {
...
- * @param {InternalJob} job
- * @return {Promise<object>}
  */
-exports.enqueueJob = async job => {
+export async function enqueueJob(job: InternalJob): Promise<object> {
   // When we already have a job that's executing, return the same promise.
   // we have another check in our createJobV2 action to return jobs that have been done in a previous gatsby run
   if (jobsInProcess.has(job.contentDigest)) {
-    return jobsInProcess.get(job.contentDigest).deferred.promise
+    return jobsInProcess.get(job.contentDigest)!.deferred.promise
```
Comment (pvdz) [inline packages/gatsby/src/utils/jobs-manager.ts:278]: Since jobs are deleted from the map, wouldn't it be _possible_ for this to be `undefined`?

File: packages/gatsby/src/utils/jobs-manager.ts (hunk 4 above)
Comment (pvdz) [reply packages/gatsby/src/utils/jobs-manager.ts:91]: (We can do this in a followup)

**Comment by pvdz** [comment]
Ah looks like `WorkerError` has to go above the types (because classes aren't hoisted and TS won't do this either). Sorry about that.

File: packages/gatsby/src/utils/jobs-manager.ts (hunk 6 above)
Comment (martijnjanssen) [reply packages/gatsby/src/utils/jobs-manager.ts]: Good spot, I think I missed that. I can't think of a good reason why that would've been removed.

File: packages/gatsby/src/utils/jobs-manager.ts (hunk 5 above)
Comment (martijnjanssen) [reply packages/gatsby/src/utils/jobs-manager.ts:33]: I think that would be a nice change to make, just had a look at it, but I wasn't able to make that change quickly. I think that would be nice to do in an entirely different PR.

**Comment by martijnjanssen** [comment]
Looks like my editor removed all trailing `,`'s... I'm adding them back now.

**Comment by martijnjanssen** [comment]
Okay there are some pre-commit tasks which keep formatting the file and removing trailing `,`'s from the file, would you know anything about that @pvdz?

**Comment by pvdz** [comment]
I believe it's caused by linting? What happens when you do `yarn format`? (or `npm run format`)
image

I'll proc the formatting from a bot (this will cause remote changes for you. When necessary, either pull before pushing or force push)

**Comment by pvdz** [comment]
I'm actually about to go on vacation so if we can't resolve that before then, somebody else will take over. But I think this is about ready to merge.

**Comment by pvdz** [comment]
This seems like an actual problem:
> ../gatsby/src/redux/types.ts(6,10): error TS2305: Module '"../utils/jobs-manager"' has no exported member 'InternalJobInterface'.

**Comment by pvdz** [comment]
@martijnjanssen friendly ping :)

**Comment by pvdz** [comment]
I've went ahead and fixed the two remaining typing problems.
